### ISSUE TITLE
`setTimeout` is not running as expected when using `prompt()`

### ISSUE THREAD

**Issue by Hexagon** [denoland/deno#24053]
Version: Deno 1.44

setTimeout is not running as expected when using prompt() to collect input from the user. This makes it impossible to do background work during a series of user inputs.

```typescript
const startTime = Date.now();

// Set up a timeout loop executing once every second
let executions = -1; // recursiveTimeout will fire once on the initial invocation, start at -1 to leave it as 0.
function recursiveTimeout() {
    executions++;
    setTimeout(()=>{recursiveTimeout()}, 1000);
}
recursiveTimeout();

// ---
function printStatus() {
    console.log(`Recursive timeout has recursed ${executions} times in ${Math.round(((Date.now()-startTime)/1000))} seconds.`);
}

// Gather three inputs from the user
console.log("Hello from main");
printStatus();
const _promptResult1 = prompt("Say something: ");
// The timeout should have timed out, and should be first in queue to execute now
// - but excecutions is still 0
printStatus();
// it should definitely have fired now, i think...
const _promptResult2 = prompt("Say something more: ");
// Same again ...
printStatus();
const _promptResult3 = prompt("Say something more: ");
// Still zero?
printStatus();

console.log("Ok bye!");
Deno.exit();
```

Gives

```
Hello from main
Recursive timeout has recursed 0 times in 0 seconds.
Say something:  5
Recursive timeout has recursed 0 times in 3 seconds.
Say something more:  hello
Recursive timeout has recursed 0 times in 8 seconds.
Say something more:  what?
Recursive timeout has recursed 0 times in 11 seconds.
Ok bye!
```

**Comment by Hexagon** [comment]
Throwing in a dummy timeout between each prompt seem to resolve the issue

```typescript
const startTime = Date.now();

// Set up a timeout loop executing once every second
let executions = -1;
function recursiveTimeout() {
    executions++;
    setTimeout(()=>{recursiveTimeout()}, 1000);
}
recursiveTimeout();

// ---
function printStatus() {
    console.log(`Recursive timeout has recursed ${executions} times in ${Math.round(((Date.now()-startTime)/1000))} seconds.`);
}


// Gather three inputs from the user
console.log("Hello from main");
printStatus();
const _promptResult1 = prompt("Say something: ");
// The timeout should have timed out, and should be first in queue to execute now
// - but excecutions is still 0
printStatus();
await new Promise(resolve => setTimeout(resolve, 0));
const _promptResult2 = prompt("Say something more: ");
// Same again ...
printStatus();
await new Promise(resolve => setTimeout(resolve, 0));
const _promptResult3 = prompt("Say something more: ");
// Still zero?
printStatus();

console.log("Ok bye!");
Deno.exit();
```

```
Hello from main
Recursive timeout has recursed 0 times in 0 seconds.
Say something:  asdasdasd
Recursive timeout has recursed 0 times in 1 seconds.
Say something more:  asdasdasd
Recursive timeout has recursed 1 times in 3 seconds.
Say something more:  asdasdsad
Recursive timeout has recursed 2 times in 6 seconds.
Ok bye!
```

**Comment by dsherret** [comment]
Here's an example with an async read:

```ts
import $ from "jsr:@david/dax@0.41.0";
const startTime = Date.now();

// Set up a timeout loop executing once every second
let executions = -1; // recursiveTimeout will fire once on the initial invocation, start at -1 to leave it as 0.
function recursiveTimeout() {
  executions++;
  setTimeout(() => {
    recursiveTimeout();
  }, 1000);
}
recursiveTimeout();

// ---
function printStatus() {
  console.log(
    `Recursive timeout has recursed ${executions} times in ${Math.round((Date.now() - startTime) / 1000)} seconds.`,
  );
}

// Gather three inputs from the user
console.log("Hello from main");
printStatus();
const _promptResult1 = await $.prompt("Say something: ");
// The timeout should have timed out, and should be first in queue to execute now
// - but excecutions is still 0
printStatus();
// it should definitely have fired now, i think...
const _promptResult2 = await $.prompt("Say something more: ");
// Same again ...
printStatus();
const _promptResult3 = await $.prompt("Say something more: ");
// Still zero?
printStatus();

console.log("Ok bye!");
```

**Comment by Hexagon** [comment]
Nice solution, i'll use that instead!

But shouldn't the event loop finish some time during

```
// Gather three inputs from the user
console.log("Hello from main");
printStatus();
const _promptResult1 = prompt("Say something: ");
// The timeout should have timed out, and should be first in queue to execute now
// - but excecutions is still 0
printStatus();
await new Promise(resolve => setTimeout(resolve, 0));
const _promptResult2 = prompt("Say something more: ");
// Same again ...
printStatus();
await new Promise(resolve => setTimeout(resolve, 0));
const _promptResult3 = prompt("Say something more: ");
// Still zero?
printStatus();
```

In my real world scenario, lots and lots of work is done, but the setTimeout never fires if prompt is in the loop. If the prompt is omitted, but the same work is done, setTimeout execute just fine. It seems like prompt blocks other operations more intrusively than normal blocking operations.

**Comment by dsherret** [comment]
> But shouldn't the event loop finish some time during

No, because execution is blocked by a sychronous call.

> It seems like prompt blocks other operations more intrusively than normal blocking operations.

Can you give an example? If you do any synchronous blocking call in javascript then a `setTimeout` will be unable to be executed because JS execution is blocked.

>  and yielding to the event loop by adding `await new Promise(resolve => setTimeout(resolve, 0));`

Yielding like this should be a bit more efficient:

```ts
await new Promise(resolve => resolve());
```

**Comment by Hexagon** [comment]
Thanks! I realize my example is flawed. Tried to make a more accurate reproduction but can not get it to work. Closing until i eventually find a way to reproduce it 👍

**Comment by BlackAsLight** [comment]
To give more info on why setTimeout never executed, JavaScript is single threaded and will only work on one thing at a time.

The event loop will only give the stack a new task to work on when the stack finishes what it is doing, or some code (like where an await keyword is) forces said current task to the back of the event loop. JavaScript won't force sync code to the back of the event loop.

Functions like prompt are considered event blocking because it forces the stack to wait for external events to happen before continuing on with the execution, instead of pushing the task to the back of the event loop so it can work on something else in the meantime. 

An empty while loop true will also block the event loop creating a freezing, non-responsive effect as there is no where inside that loop that forces the stack to push it to the back of the event loop. 
```ts
setTimeout(() => console.log(1), 0)
while (true) {}
```
This is also why it's recommended to put computational heavy tasks in workers, so the user doesn't experience a type of lagging, as workers have their own thread. 

It should be noted though that JavaScript tries to run async code in a synchronous manner, and will only push the current task to the back of the event loop when it is forced to do so. Meaning the mere presence of async await syntax does not mean it will switch tasks.
